# Limpieza de los datos

In [1]:
import pandas as pd

In [2]:
# aquí están los datos en local
data_path = '/Users/cbautistap/Dropbox/MCD/Cursos/primavera-2021/dpa/data/Food_Inspections.csv'

In [3]:
# cargamos datos, convertimos a formato datetime la columna 'Inspection Date' y queda con formato yy/mm/dd
data = pd.read_csv(data_path, parse_dates=['Inspection Date'])

In [4]:
data.shape

(215130, 17)

In [5]:
#data['month'] = data['Inspection Date'].dt.month

In [6]:
# creamos función para estandarizar el nombre de las columnas.
def standarize_column_names(data, excluded_punctuation=".,-*¿?¡!#"):
    data.columns = data.columns.str.lower().str.replace(" ", "_")
    for ch in excluded_punctuation:
        data.columns = data.columns.str.replace(ch, "")

In [7]:
standarize_column_names(data)

/Users/cbautistap/.pyenv/versions/itam_dpa/lib/python3.7/site-packages/ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  """


In [8]:
# cambiamos el nombre de la columna license
data=data.rename(columns={'license_':'license'})
data.head(2)

inspection_id      dba_name      aka_name    license facility_type  \
0        2472391  PHILZ COFFEE  PHILZ COFFEE  2766979.0           NaN   
1        2453551          BIAN          BIAN  2744753.0    Restaurant   

              risk             address     city state      zip  \
0  Risk 2 (Medium)    1555 N WELLS ST   CHICAGO    IL  60610.0   
1    Risk 1 (High)  600 W CHICAGO AVE   CHICAGO    IL  60654.0   

  inspection_date inspection_type    results violations   latitude  longitude  \
0      2021-01-07         License  Not Ready        NaN  41.910736 -87.634551   
1      2020-10-20         License       Pass        NaN  41.896585 -87.642996   

                                   location  
0  (-87.63455069145124, 41.910736031820235)  
1  (-87.64299618172501, 41.896585191199556)

In [9]:
data.dtypes

inspection_id               int64
dba_name                   object
aka_name                   object
license                   float64
facility_type              object
risk                       object
address                    object
city                       object
state                      object
zip                       float64
inspection_date    datetime64[ns]
inspection_type            object
results                    object
violations                 object
latitude                  float64
longitude                 float64
location                   object
dtype: object

Queremos convertir license y zip a string. Si lo convertimos directo astype(str) queda con decimales, por lo tanto convertimos primero a formato int. Sin embargo, ambas columnas tienen NAs por lo que una solución es llenarlos con 0 y ahora sí convertir a string. QUIZÁ NO HAY PROBLEMA TENERLOS EN STRING CON EL DECIMAL XXXX.0

In [10]:
data.isna().sum()

inspection_id          0
dba_name               0
aka_name            2487
license               17
facility_type       4892
risk                  69
address                0
city                 158
state                 50
zip                   52
inspection_date        0
inspection_type        1
results                0
violations         57447
latitude             710
longitude            710
location             710
dtype: int64

In [11]:
#Subset de data frame con missing values en cualquier columna
null_data = data[data.isnull().any(axis=1)]
null_data.head()

inspection_id       dba_name       aka_name    license facility_type  \
0        2472391   PHILZ COFFEE   PHILZ COFFEE  2766979.0           NaN   
1        2453551           BIAN           BIAN  2744753.0    Restaurant   
2        2386633         JIN JU         JIN JU    27137.0    Restaurant   
3        2386595  LA BIZNAGA #2  LA BIZNAGA #2  2708992.0           NaN   
4        2386523    MC DONALD'S    MC DONALD'S  2560785.0    Restaurant   

              risk              address     city state      zip  \
0  Risk 2 (Medium)     1555 N WELLS ST   CHICAGO    IL  60610.0   
1    Risk 1 (High)   600 W CHICAGO AVE   CHICAGO    IL  60654.0   
2    Risk 1 (High)     5203 N CLARK ST   CHICAGO    IL  60640.0   
3    Risk 1 (High)  2949 W BELMONT AVE   CHICAGO    IL  60618.0   
4  Risk 2 (Medium)  6336 S ASHLAND AVE   CHICAGO    IL  60636.0   

  inspection_date       inspection_type    results violations   latitude  \
0      2021-01-07               License  Not Ready        NaN  41.910736   
1      2020-10-20               License       Pass        NaN  41.896585   
2      2020-08-28               Canvass   No Entry        NaN  41.976301   
3      2020-08-27             Complaint   No Entry        NaN  41.939256   
4      2020-08-26  Short Form Complaint       Pass        NaN  41.778361   

   longitude                                  location  
0 -87.634551  (-87.63455069145124, 41.910736031820235)  
1 -87.642996  (-87.64299618172501, 41.896585191199556)  
2 -87.668276   (-87.66827593789948, 41.97630115368914)  
3 -87.702270  (-87.70226967930802, 41.939255926667535)  
4 -87.664337   (-87.66433686342792, 41.77836097603761)

### Queremos entender qué tipo de observaciones son las que tienen missing values, con el objetivo de encontrar algún patrón o algo que nos permita imputar de mejor manera. 

Para eso, haremos subset de missing values por columna. Empezamos con las que tienen mayor NaN.

Nota. La columna de aka_name no la exploraremos. Es lógico que haya mv para esta columna, además, no la utilizaremos para el modelo

In [12]:
#queremos ver que observaciones son las que tienen missing values en la columna 'violations'
null_violations = null_data[null_data.violations.isnull()]
null_violations.shape

(57447, 17)

In [15]:
#Missing values en la columna 'violations'
null_data[null_data.violations.isnull()].groupby('results').count()

inspection_id  dba_name  aka_name  license  \
results                                                            
Business Not Located             72        72        57       72   
Fail                           3183      3183      2989     3182   
No Entry                       6796      6796      6709     6796   
Not Ready                      2226      2226      2172     2226   
Out of Business               18677     18677     18098    18676   
Pass                          25925     25925     25491    25922   
Pass w/ Conditions              568       568       563      568   

                      facility_type   risk  address   city  state    zip  \
results                                                                    
Business Not Located             45     64       72     70     72     70   
Fail                           2842   3162     3183   3170   3183   3178   
No Entry                       6725   6789     6796   6790   6793   6795   
Not Ready                      2123   2210     2226   2212   2222   2223   
Out of Business               14638  18673    18677  18658  18674  18663   
Pass                          25729  25914    25925  25895  25913  25916   
Pass w/ Conditions              562    568      568    568    567    568   

                      inspection_date  inspection_type  violations  latitude  \
results                                                                        
Business Not Located               72               72           0        69   
Fail                             3183             3183           0      3168   
No Entry                         6796             6796           0      6794   
Not Ready                        2226             2226           0      2218   
Out of Business                 18677            18677           0     18641   
Pass                            25925            25924           0     25799   
Pass w/ Conditions                568              568           0       567   

                      longitude  location  
results                                    
Business Not Located         69        69  
Fail                       3168      3168  
No Entry                   6794      6794  
Not Ready                  2218      2218  
Out of Business           18641     18641  
Pass                      25799     25799  
Pass w/ Conditions          567       567

Agrupando por Results, el 89.4% de los NaN de la columna 'Violations' están en 'Pass', 'Out of Business' y 'No Entry'. Tiene sentido que los que pasaron la inspección no tengan violaciones, asimismo, que los que ya estén cerrados y en los cuales no se entró, no tengan violaciones.

Por otro lado, solo el 5% de los mv de la columna 'Violations' son de resultados 'Fail' (que representan 1.47% del total de observaciones).
Dentro de éstas, el 62% son de Tipo de Inspección 'License'

In [16]:
results_fail = null_violations['results']=='Fail'

In [18]:
null_violations[results_fail].groupby('inspection_type').count()

inspection_id  dba_name  aka_name  \
inspection_type                                                               
Business Not Located                                  1         1         1   
Canvass                                             462       462       453   
Canvass Re-Inspection                                16        16        16   
Complaint                                           128       128       118   
Complaint Re-Inspection                               6         6         6   
Complaint-Fire                                       22        22        22   
Consultation                                         45        45        42   
Duplicated                                            1         1         1   
Illegal Operation                                     1         1         0   
LICENSE REQUEST                                       4         4         4   
LICENSE WRONG ADDRESS                                 1         1         0   
LICENSE/NOT READY                                     2         2         2   
LIQOUR TASK FORCE NOT READY                           1         1         1   
License                                            1999      1999      1850   
License Re-Inspection                                65        65        57   
License-Task Force                                  100       100        98   
NO ENTRY                                              6         6         6   
NO ENTRY-SHORT COMPLAINT)                             1         1         1   
No Entry                                             28        28        28   
Non-Inspection                                        5         5         5   
Not Ready                                             6         6         6   
OUT OF BUSINESS                                       8         8         7   
OWNER SUSPENDED OPERATION/LICENSE                     1         1         0   
Out of Business                                     112       112       110   
Package Liquor 1474                                   7         7         7   
Pre-License Consultation                             11        11         9   
RECALL INSPECTION                                     1         1         1   
Recent Inspection                                     5         5         5   
Short Form Complaint                                 57        57        55   
Short Form Fire-Complaint                            17        17        17   
Special Events (Festivals)                            2         2         2   
Suspected Food Poisoning                              3         3         3   
TASK FORCE LIQUOR 1470                                1         1         1   
TASK FORCE NOT READY                                  1         1         1   
Tag Removal                                           3         3         3   
Task Force Liquor 1475                               47        47        44   
Task Force for liquor 1474                            1         1         1   
finish complaint inspection from 5-18-10              1         1         1   
no entry                                              4         4         4   
out ofbusiness                                        1         1         1   

                                          license  facility_type  risk  \
inspection_type                                                          
Business Not Located                            1              1     1   
Canvass                                       462            340   462   
Canvass Re-Inspection                          16             16    16   
Complaint                                     128            117   125   
Complaint Re-Inspection                         6              6     6   
Complaint-Fire                                 22             20    22   
Consultation                                   45             41    43   
Duplicated                                      1        

In [30]:
# MV de columna 'Facility Type'
null_data[null_data.facility_type.isnull()].groupby('results').count()

inspection_id  dba_name  aka_name  license  \
results                                                            
Business Not Located             27        27        15       27   
Fail                            368       368       310      368   
No Entry                         71        71        54       71   
Not Ready                       103       103        72      103   
Out of Business                4040      4040      3996     4040   
Pass                            245       245       235      245   
Pass w/ Conditions               38        38        37       38   

                      facility_type  risk  address  city  state   zip  \
results                                                                 
Business Not Located              0    19       27    25     27    25   
Fail                              0   354      368   367    368   367   
No Entry                          0    64       71    71     71    71   
Not Ready                         0    87      103   100    103   103   
Out of Business                   0  4036     4040  4039   4040  4040   
Pass                              0   240      245   243    245   245   
Pass w/ Conditions                0    38       38    38     38    38   

                      inspection_date  inspection_type  violations  latitude  \
results                                                                        
Business Not Located               27               27           0        24   
Fail                              368              368          27       367   
No Entry                           71               71           0        71   
Not Ready                         103              103           0       103   
Out of Business                  4040             4040           1      4034   
Pass                              245              245          49       243   
Pass w/ Conditions                 38               38          32        36   

                      longitude  location  
results                                    
Business Not Located         24        24  
Fail                        367       367  
No Entry                     71        71  
Not Ready                   103       103  
Out of Business            4034      4034  
Pass                        243       243  
Pass w/ Conditions           36        36

El 84% de los NaN en la columna de Facility Type corresponden a results 'Out of business', 'Business Not Located' y 'No Entry'

In [32]:
# MV de columna 'location'
null_data[null_data.location.isnull()].groupby('results').count()

inspection_id  dba_name  aka_name  license  \
results                                                            
Business Not Located              3         3         1        3   
Fail                            145       145       143      145   
No Entry                          3         3         3        3   
Not Ready                         8         8         8        8   
Out of Business                  36        36        35       36   
Pass                            432       432       426      432   
Pass w/ Conditions               83        83        82       83   

                      facility_type  risk  address  city  state  zip  \
results                                                                
Business Not Located              0     1        3     3      3    3   
Fail                            144   145      145   145    145  145   
No Entry                          3     3        3     3      3    3   
Not Ready                         8     8        8     7      8    7   
Out of Business                  30    36       36    35     36   35   
Pass                            430   432      432   431    432  431   
Pass w/ Conditions               81    83       83    83     83   83   

                      inspection_date  inspection_type  violations  latitude  \
results                                                                        
Business Not Located                3                3           0         0   
Fail                              145              145         130         0   
No Entry                            3                3           1         0   
Not Ready                           8                8           0         0   
Out of Business                    36               36           0         0   
Pass                              432              432         306         0   
Pass w/ Conditions                 83               83          82         0   

                      longitude  location  
results                                    
Business Not Located          0         0  
Fail                          0         0  
No Entry                      0         0  
Not Ready                     0         0  
Out of Business               0         0  
Pass                          0         0  
Pass w/ Conditions            0         0

En total son 710 observaciones con mv en ubicación (que coinciden con coordenadas geográficas). La mayoría son de establecimientos que pasaron la visita. Asimismo, destaca que por facility type, la mayoría son de tipo school y también incluye algunos establecimientos móviles.

In [33]:
# MV de columna 'location', agrupando con columna 'facility_type'
null_data[null_data.location.isnull()].groupby('facility_type').count()

inspection_id  dba_name  aka_name  license  \
facility_type                                                                  
BREWERY                                       4         4         4        4   
CHURCH/SPECIAL EVENT                          3         3         0        3   
Daycare Combo 1586                            1         1         0        1   
Food Vending Machines                         1         1         1        1   
GAS STATION                                   1         1         1        1   
Golden Diner                                  9         9         9        9   
Grocery Store                                25        25        25       25   
HIGH SCHOOL KITCHEN                           3         3         3        3   
Hospital                                     10        10        10       10   
Long Term Care                                7         7         7        7   
MFD TRUCK                                     5         5         5        5   
Mobile Food Dispenser                        75        75        70       75   
Mobile Food Preparer                         43        43        43       43   
Mobile Frozen Desserts Vendor                 9         9         9        9   
Mobile Prepared Food Vendor                   5         5         5        5   
PUSHCART                                      2         2         2        2   
Restaurant                                  136       136       136      136   
STORE                                         1         1         0        1   
School                                      320       320       320      320   
Shared Kitchen User (Long Term)              30        30        30       30   
Shared Kitchen User (Short Term)              4         4         4        4   
Special Event                                 2         2         2        2   

                                  risk  address  city  state  zip  \
facility_type                                                       
BREWERY                              4        4     4      4    4   
CHURCH/SPECIAL EVENT                 3        3     3      3    3   
Daycare Combo 1586                   1        1     1      1    1   
Food Vending Machines                1        1     1      1    1   
GAS STATION                          1        1     1      1    1   
Golden Diner                         9        9     8      9    8   
Grocery Store                       25       25    25     25   25   
HIGH SCHOOL KITCHEN                  3        3     3      3    3   
Hospital                            10       10    10     10   10   
Long Term Care                       7        7     7      7    7   
MFD TRUCK                            5        5     5      5    5   
Mobile Food Dispenser               75       75    75     75   75   
Mobile Food Preparer                43       43    43     43   43   
Mobile Frozen Desserts Vendor        9        9     9      9    9   
Mobile Prepared Food Vendor          5        5     5      5    5   
PUSHCART                             2        2     2      2    2   
Restaurant                         136      136   134    136  134   
STORE                                1        1     1      1    1   
School                             320      320   320    320  320   
Shared Kitchen User (Long Term)     30       30    30     30   30   
Shared Kitchen User (Short Term)     4        4     4      4    4   
Special Event                        2        2     2      2    2   

                                  inspection_date  inspection_type  results  \
facility_type                                                                 
BREWERY                                         4                4        4   
CHURCH/SPECIAL EVENT                            3                3        3   
Daycare Combo 1586                              1                1        1   
Food Vending Machines                           1            

In [37]:
# MV de columna 'risk', agrupando por 'results'
null_data[null_data.risk.isnull()].groupby('results').count()

inspection_id  dba_name  aka_name  license  \
results                                                            
Business Not Located              8         8         2        8   
Fail                             22        22        15       22   
No Entry                          7         7         2        7   
Not Ready                        16        16         9       16   
Out of Business                   4         4         3        4   
Pass                             12        12         8       12   

                      facility_type  risk  address  city  state  zip  \
results                                                                
Business Not Located              0     0        8     8      8    8   
Fail                              8     0       22    22     22   22   
No Entry                          0     0        7     7      7    7   
Not Ready                         0     0       16    14     16   16   
Out of Business                   0     0        4     4      4    4   
Pass                              7     0       12    11     12   12   

                      inspection_date  inspection_type  violations  latitude  \
results                                                                        
Business Not Located                8                8           0         6   
Fail                               22               22           1        22   
No Entry                            7                7           0         7   
Not Ready                          16               16           0        16   
Out of Business                     4                4           0         4   
Pass                               12               12           1        12   

                      longitude  location  
results                                    
Business Not Located          6         6  
Fail                         22        22  
No Entry                      7         7  
Not Ready                    16        16  
Out of Business               4         4  
Pass                         12        12

In [40]:
# MV de columna 'risk', agrupando por 'inspection_type'
null_data[null_data.risk.isnull()].groupby('inspection_type').count()

inspection_id  dba_name  aka_name  license  \
inspection_type                                                          
Canvass                                 3         3         2        3   
Complaint                              10        10         4       10   
Complaint-Fire                          1         1         0        1   
Consultation                            5         5         4        5   
Illegal Operation                       1         1         0        1   
License                                36        36        20       36   
Non-Inspection                          1         1         1        1   
Out of Business                         1         1         1        1   
Pre-License Consultation                3         3         2        3   
Short Form Complaint                    7         7         4        7   
Special Events (Festivals)              1         1         1        1   

                            facility_type  risk  address  city  state  zip  \
inspection_type                                                              
Canvass                                 0     0        3     3      3    3   
Complaint                               3     0       10    10     10   10   
Complaint-Fire                          0     0        1     1      1    1   
Consultation                            4     0        5     5      5    5   
Illegal Operation                       1     0        1     1      1    1   
License                                 2     0       36    34     36   36   
Non-Inspection                          0     0        1     1      1    1   
Out of Business                         0     0        1     1      1    1   
Pre-License Consultation                2     0        3     3      3    3   
Short Form Complaint                    3     0        7     6      7    7   
Special Events (Festivals)              0     0        1     1      1    1   

                            inspection_date  results  violations  latitude  \
inspection_type                                                              
Canvass                                   3        3           0         3   
Complaint                                10       10           1         9   
Complaint-Fire                            1        1           0         1   
Consultation                              5        5           1         5   
Illegal Operation                         1        1           0         1   
License                                  36       36           0        35   
Non-Inspection                            1        1           0         1   
Out of Business                           1        1           0         1   
Pre-License Consultation                  3        3           0         3   
Short Form Complaint                      7        7           0         7   
Special Events (Festivals)                1        1           0         1   

                            longitude  location  
inspection_type                                  
Canvass                             3         3  
Complaint                           9         9  
Complaint-Fire                      1         1  
Consultation                        5         5  
Illegal Operation                   1         1  
License                            35        35  
Non-Inspection                      1         1  
Out of Business                     1         1  
Pre-License Consultation            3         3  
Short Form Complaint                7         7  
Special Events (Festivals)          1         1

## Por lo pronto, para proseguir con limpieza de datos, se decide reemplazar los MV con 0 para variables numéricas y con el valor 'na' para variables de texto.

### En primer lugar, convertimos a variables string las de license y zip 

In [43]:
# Reemplazamos con 0 los mv en estas dos columnas
data['license']=data['license'].fillna(0)
data['zip']=data['zip'].fillna(0)

In [44]:
data.isna().sum()

inspection_id          0
dba_name               0
aka_name            2487
license                0
facility_type       4892
risk                  69
address                0
city                 158
state                 50
zip                    0
inspection_date        0
inspection_type        1
results                0
violations         57447
latitude             710
longitude            710
location             710
dtype: int64

In [46]:
# convertimos columnas 'license' y 'zip'. Primero a int y luego a str
data = data.astype({"license": 'int', "zip": 'int'})
data = data.astype({"license": 'str', "zip": 'str'})
data.dtypes

inspection_id               int64
dba_name                   object
aka_name                   object
license                    object
facility_type              object
risk                       object
address                    object
city                       object
state                      object
zip                        object
inspection_date    datetime64[ns]
inspection_type            object
results                    object
violations                 object
latitude                  float64
longitude                 float64
location                   object
dtype: object

In [48]:
data.head(2)

inspection_id      dba_name      aka_name  license facility_type  \
0        2472391  PHILZ COFFEE  PHILZ COFFEE  2766979           NaN   
1        2453551          BIAN          BIAN  2744753    Restaurant   

              risk             address     city state    zip inspection_date  \
0  Risk 2 (Medium)    1555 N WELLS ST   CHICAGO    IL  60610      2021-01-07   
1    Risk 1 (High)  600 W CHICAGO AVE   CHICAGO    IL  60654      2020-10-20   

  inspection_type    results violations   latitude  longitude  \
0         License  Not Ready        NaN  41.910736 -87.634551   
1         License       Pass        NaN  41.896585 -87.642996   

                                   location  
0  (-87.63455069145124, 41.910736031820235)  
1  (-87.64299618172501, 41.896585191199556)

In [53]:
# Hacemos subset de las columnas que utilizaremos para el modelo
data = data[['inspection_id', 'facility_type', 'risk', 'zip', 'inspection_date','inspection_type', 'results', 'violations', 'latitude', 'longitude']]
data.head()

inspection_id facility_type             risk    zip inspection_date  \
0        2472391           NaN  Risk 2 (Medium)  60610      2021-01-07   
1        2453551    Restaurant    Risk 1 (High)  60654      2020-10-20   
2        2386633    Restaurant    Risk 1 (High)  60640      2020-08-28   
3        2386595           NaN    Risk 1 (High)  60618      2020-08-27   
4        2386523    Restaurant  Risk 2 (Medium)  60636      2020-08-26   

        inspection_type   results violations   latitude  longitude  
0               License  Not Pass        NaN  41.910736 -87.634551  
1               License      Pass        NaN  41.896585 -87.642996  
2               Canvass  Not Pass        NaN  41.976301 -87.668276  
3             Complaint  Not Pass        NaN  41.939256 -87.702270  
4  Short Form Complaint      Pass        NaN  41.778361 -87.664337

In [54]:
data.isnull().sum()

inspection_id          0
facility_type       4892
risk                  69
zip                    0
inspection_date        0
inspection_type        1
results                0
violations         57447
latitude             710
longitude            710
dtype: int64

In [55]:
data[['facility_type', 'risk', 'inspection_type','violations']]=data[['facility_type', 'risk', 'inspection_type','violations']].fillna('na')

In [56]:
data.isnull().sum()

inspection_id        0
facility_type        0
risk                 0
zip                  0
inspection_date      0
inspection_type      0
results              0
violations           0
latitude           710
longitude          710
dtype: int64

In [58]:
# Cambiamos valor de results a 'Pass' y 'Not Pass' para cualquier otro
data['results'].mask(data['results'] != 'Pass', other = 'Not Pass', inplace=True)

In [59]:
data.groupby('results').count()

inspection_id  facility_type    risk     zip  inspection_date  \
results                                                                   
Not Pass         102489         102489  102489  102489           102489   
Pass             112641         112641  112641  112641           112641   

          inspection_type  violations  latitude  longitude  
results                                                     
Not Pass           102489      102489    102211     102211  
Pass               112641      112641    112209     112209

Ahora, estandarizamos los valores de las columnas (quitamos acentos y eliminamos caracteres especiales)

In [60]:
def standarize_column_strings(data, columns, excluded_punctuation=".,-*¿?¡!"):
    for col in columns:
        data[col] = data[col].apply(lambda x: x.lower().replace(" ", "_"))
        data[col] = data[col].apply(lambda x: x.lower().replace("á", "a"))
        data[col] = data[col].apply(lambda x: x.lower().replace("é", "e"))
        data[col] = data[col].apply(lambda x: x.lower().replace("í", "i"))
        data[col] = data[col].apply(lambda x: x.lower().replace("ó", "o"))
        data[col] = data[col].apply(lambda x: x.lower().replace("ú", "u"))
        data[col] = data[col].apply(lambda x: x.lower().replace("ü", "u"))
        for ch in excluded_punctuation:
            data[col] = data[col].apply(lambda x: x.replace(ch, ""))

Tengo que hacerlo primero con las columnas que no tienen NAs

In [53]:
data.dtypes

inspection_id               int64
facility_type              object
risk                       object
zip                        object
inspection_date    datetime64[ns]
inspection_type            object
results                    object
violations                 object
latitude                  float64
longitude                 float64
dtype: object

In [55]:
col_text_nona = ['zip','results']
standarize_column_strings(data, col_text_nona)
data.head()

inspection_id facility_type             risk    zip inspection_date  \
0        2472391           NaN  Risk 2 (Medium)  60610      2021-01-07   
1        2453551    Restaurant    Risk 1 (High)  60654      2020-10-20   
2        2386633    Restaurant    Risk 1 (High)  60640      2020-08-28   
3        2386595           NaN    Risk 1 (High)  60618      2020-08-27   
4        2386523    Restaurant  Risk 2 (Medium)  60636      2020-08-26   

        inspection_type   results violations   latitude  longitude  
0               License  not_pass        NaN  41.910736 -87.634551  
1               License      pass        NaN  41.896585 -87.642996  
2               Canvass  not_pass        NaN  41.976301 -87.668276  
3             Complaint  not_pass        NaN  41.939256 -87.702270  
4  Short Form Complaint      pass        NaN  41.778361 -87.664337

In [59]:
data[['facility_type','risk','inspection_type','violations']]=data[['facility_type','risk','inspection_type','violations']].fillna('NA')
#data['risk']=data['Zip'].fillna(0)

In [60]:
data.isna().sum()

inspection_id          0
facility_type          0
risk                   0
zip                    0
inspection_date        0
inspection_type        1
results                0
violations         57447
latitude             710
longitude            710
dtype: int64

In [61]:
data['violations'].isna()

0          True
1          True
2          True
3          True
4          True
          ...  
215125    False
215126     True
215127    False
215128    False
215129     True
Name: violations, Length: 215130, dtype: bool

inspection_id                    facility_type             risk  \
0             2472391                               NA  Risk 2 (Medium)   
1             2453551                       Restaurant    Risk 1 (High)   
2             2386633                       Restaurant    Risk 1 (High)   
3             2386595                               NA    Risk 1 (High)   
4             2386523                       Restaurant  Risk 2 (Medium)   
...               ...                              ...              ...   
215116         176497                               NA    Risk 1 (High)   
215121          67785                       Restaurant  Risk 2 (Medium)   
215122         120351                        Wholesale  Risk 2 (Medium)   
215126         215221                           School    Risk 1 (High)   
215129          67853  Daycare Above and Under 2 Years    Risk 1 (High)   

          zip inspection_date        inspection_type   results violations  \
0       60610      2021-01-07                License  not_pass        NaN   
1       60654      2020-10-20                License      pass        NaN   
2       60640      2020-08-28                Canvass  not_pass        NaN   
3       60618      2020-08-27              Complaint  not_pass        NaN   
4       60636      2020-08-26   Short Form Complaint      pass        NaN   
...       ...             ...                    ...       ...        ...   
215116  60657      2010-03-04        Out of Business  not_pass        NaN   
215121  60629      2010-01-07  License Re-Inspection      pass        NaN   
215122  60612      2010-02-22  License Re-Inspection      pass        NaN   
215126  60638      2010-02-19                Canvass      pass        NaN   
215129  60629      2010-01-12                Canvass      pass        NaN   

         latitude  longitude  
0       41.910736 -87.634551  
1       41.896585 -87.642996  
2       41.976301 -87.668276  
3       41.939256 -87.702270  
4       41.778361 -87.664337  
...           ...        ...  
215116  41.934436 -87.649220  
215121  41.779225 -87.693456  
215122  41.886874 -87.677296  
215126  41.775492 -87.758769  
215129  41.778853 -87.703781  

[57966 rows x 10 columns]

inspection_id                    facility_type             risk  \
0             2472391                               NA  Risk 2 (Medium)   
1             2453551                       Restaurant    Risk 1 (High)   
2             2386633                       Restaurant    Risk 1 (High)   
3             2386595                               NA    Risk 1 (High)   
4             2386523                       Restaurant  Risk 2 (Medium)   
...               ...                              ...              ...   
215116         176497                               NA    Risk 1 (High)   
215121          67785                       Restaurant  Risk 2 (Medium)   
215122         120351                        Wholesale  Risk 2 (Medium)   
215126         215221                           School    Risk 1 (High)   
215129          67853  Daycare Above and Under 2 Years    Risk 1 (High)   

          zip inspection_date        inspection_type   results violations  \
0       60610      2021-01-07                License  not_pass        NaN   
1       60654      2020-10-20                License      pass        NaN   
2       60640      2020-08-28                Canvass  not_pass        NaN   
3       60618      2020-08-27              Complaint  not_pass        NaN   
4       60636      2020-08-26   Short Form Complaint      pass        NaN   
...       ...             ...                    ...       ...        ...   
215116  60657      2010-03-04        Out of Business  not_pass        NaN   
215121  60629      2010-01-07  License Re-Inspection      pass        NaN   
215122  60612      2010-02-22  License Re-Inspection      pass        NaN   
215126  60638      2010-02-19                Canvass      pass        NaN   
215129  60629      2010-01-12                Canvass      pass        NaN   

         latitude  longitude  
0       41.910736 -87.634551  
1       41.896585 -87.642996  
2       41.976301 -87.668276  
3       41.939256 -87.702270  
4       41.778361 -87.664337  
...           ...        ...  
215116  41.934436 -87.649220  
215121  41.779225 -87.693456  
215122  41.886874 -87.677296  
215126  41.775492 -87.758769  
215129  41.778853 -87.703781  

[57447 rows x 10 columns]

In [68]:
data[data.iloc[:,8]=='NaN']

Empty DataFrame
Columns: [inspection_id, facility_type, risk, zip, inspection_date, inspection_type, results, violations, latitude, longitude]
Index: []

In [58]:
data['facility_type'] = data['facility_type'].lower()

AttributeError: 'Series' object has no attribute 'lower'

In [56]:
# NO SE PUEDE POR LOS NAS ¿qué tratamiento les daremos a los na (fillna o dropna o como imputar)
standarize_column_strings(data, ['facility_type'])
#

AttributeError: 'float' object has no attribute 'lower'

In [30]:
col_text = ['dba_name',
 #'aka_name'
 'license',
 'facility_type',
 #'risk',
 #'address',
 #'city',
 #'state',
 #'zip',
 #'inspection_type',
 #'results',
 #'violations',
 #'location'
           ]
#col_text = data.select_dtypes(include=['object']).columns.to_list()
#col_text = ["alcaldia", "colonia", "indice_des"]
standarize_column_strings(data, col_text)
data.head()

AttributeError: 'float' object has no attribute 'lower'

In [19]:
col_text

['dba_name',
 'aka_name',
 'license',
 'facility_type',
 'risk',
 'address',
 'city',
 'state',
 'zip',
 'inspection_type',
 'results',
 'violations',
 'location']

In [18]:
data[col_text].dtypes

dba_name           object
aka_name           object
license            object
facility_type      object
risk               object
address            object
city               object
state              object
zip                object
inspection_type    object
results            object
violations         object
location           object
dtype: object

In [ ]:
data.groupby('results').count()
#data['results'].nunique()